src/compiler/checker.ts

```diff
@@ -24877,6 +24881,18 @@ namespace ts {
             return getSpellingSuggestionForName(isString(name) ? name : idText(name), getPropertiesOfType(containingType), SymbolFlags.Value);
         }
 
+        function getSuggestedSymbolForNonexistentJSXAttribute(name: Identifier | PrivateIdentifier | string, containingType: Type): Symbol | undefined {
+            const strName = isString(name) ? name : idText(name);
+            const properties = getPropertiesOfType(containingType);
+            const defaultSuggestion = () => getSpellingSuggestionForName(strName, properties, SymbolFlags.Value);
+            const getCandidate = (expectedName: string) => properties.find(x => symbolName(x) === expectedName);
+            switch (strName) {
+                case "for": return getCandidate("htmlFor") || defaultSuggestion();
+                case "class": return getCandidate("className") || defaultSuggestion();
+            }
+            return defaultSuggestion();

```

Comment:
Just inline those functions. This isn't Haskell and I don't trust JS VMs to inline them anyway.

```suggestion
            const jsxSpecific = strName === "for" ? properties.find(x => symbolName(x) === "htmlFor")
                : strName === "class" ? properties.find(x => symbolName(x) === "className")
                : undefined;
            return jsxSpecific ?? getSpellingSuggestionForName(strName, properties, SymbolFlags.Value);
```